tests/0085-headers.cpp

```diff
@@ -0,0 +1,219 @@
+    std::string actual_value = std::string(actual_header.value());
+    std::string expected_key = expected_header.key();
+    std::string expected_value = std::string(expected_header.value());
+    if (actual_key != expected_key) {
+      Test::Fail(tostr() << "Header key does not match, expected '" 
+                 << actual_key << "' but got '" << expected_key << "'\n");
+    }
+    if (actual_value != expected_value) {
+      Test::Fail(tostr() << "Header value does not match, expected '" 
+                 << actual_value << "' but got '" << expected_value << "'\n");
+    }
+  }
+}
+
+static void test_n_headers (int n, const char* message) {
+  std::string topic = Test::mk_topic_name("0085-headers", 1);
+  RdKafka::Conf *conf;
+  std::string errstr;
+
+  Test::conf_init(&conf, NULL, 0);
+
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "socket.timeout.ms", "10000");
+  Test::conf_set(conf, "enable.auto.commit", "false");
+  Test::conf_set(conf, "enable.partition.eof", "false");
+  Test::conf_set(conf, "auto.offset.reset", "earliest");
+  Test::conf_set(conf, "statistics.interval.ms", "1000");
+
+  MyCbs cbs;
+  cbs.seen_commit = 0;
+  cbs.seen_stats  = 0;
+  if (conf->set("offset_commit_cb", (RdKafka::OffsetCommitCb *)&cbs, errstr) !=
+      RdKafka::Conf::CONF_OK)
+    Test::Fail("Failed to set commit callback: " + errstr);
+  if (conf->set("event_cb", (RdKafka::EventCb *)&cbs, errstr) !=
+      RdKafka::Conf::CONF_OK)
+    Test::Fail("Failed to set event callback: " + errstr);
+
+  RdKafka::Producer *p = RdKafka::Producer::create(conf, errstr);
+  if (!p)
+    Test::Fail("Failed to create Producer: " + errstr);
+
+  RdKafka::ErrorCode err;
+
+  std::vector<RdKafka::Headers::Header> headers_arr;
+  for (int i = 0; i < n; ++i) {
+    std::stringstream key_s;
+    key_s << "header_" << i;
+    std::string key = key_s.str();
+    std::stringstream val_s;
+    val_s << "value_" << i;
+    std::string val = val_s.str();
+    headers_arr.push_back(RdKafka::Headers::Header(key, val.c_str()));
+  }
+  RdKafka::Headers *produce_headers = RdKafka::Headers::create(headers_arr, false);
+
+  err = p->produce(topic, 0,
+                    RdKafka::Producer::RK_MSG_COPY,
+                    (void *)message, message ? strlen(message) : 0,
+                    (void *)"key", 3, 0, NULL, produce_headers);
+
+  p->flush(tmout_multip(10000));
+
+  if (p->outq_len() > 0)
+    Test::Fail(tostr() << "Expected producer to be flushed, " <<
+               p->outq_len() << " messages remain");
+
+  RdKafka::KafkaConsumer *c = RdKafka::KafkaConsumer::create(conf, errstr);
+  if (!c)
+    Test::Fail("Failed to create KafkaConsumer: " + errstr);
+  
+  std::vector<std::string> topics;
+  topics.push_back(topic);
+  if ((err = c->subscribe(topics)))
+    Test::Fail("subscribe failed: " + RdKafka::err2str(err));
+
+  int cnt = 0;
+  while (!cbs.seen_commit || !cbs.seen_stats) {
```

Comment:
these are not related to this test. Instead wait for the same number of messages you produced above.